content/en/blog/_posts/2022-11-11-retroactive-default-storage-class.md

```diff
@@ -0,0 +1,158 @@
+
+### Changing default StorageClass did not have a good solution
+
+There are two options admins had when they wanted to change default StorageClass:
+
+1. Creating the new StorageClass as default before removing the old one which results in having two defaults for a short period of time.
+If a user creates a PersistentVolumeClaim at this point with storageClassName set to <code>null</code> (which means default StorageClass) the newest default StorageClass is chosen and assigned to this PVC.
+
+2. Remove the old default first and create new default StorageClass which results in having no default for a short period of time.
+If user creates a PersistentVolumeClaim with storageClassName set to <code>nil</code> (which means default StorageClass) the PVC will be in <code>Pending</code> state forever, then user has to fix this by deleting the PVC and creating it again once default StorageClass is available.
+
+Looking at the scenarios above it is clear that neither of them is optimal.
+
+### Resource ordering during cluster installation
+
+If cluster installation tool needs to create resources that require storage (e.g. image registry) it was difficult to get the ordering right because any Pods that require storage and rely on default StorageClass presence will fail to create if there is no default at that point yet.
+
+## What changed
+
+We've changed PersistentVolume (PV) controller so it can assign a default StorageClass to any unbound PersistentVolumeClaims that have a null (unset) <code>storageClassName</code>.
+In order to allow the value change PersistentVolumeClaim admission within the API server also changed to allow changing the value from unset to an actual StorageClass name.
+
+### <code>nil</code> vs <code>""</code> for storageClassName - does it matter?
+
+Before this feature was introduced those values were equal in terms of behavior and any PersistentVolumeClaim with <code>nil</code> or <code>""</code> storageClassName value would bind to existing PersistentVolume resource with storageClassName also set to <code>nil</code> or <code>""</code>.
+
+With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name.
+With these constraints in mind the feature changes the semantics of <code>nil</code> if default StorageClass is present to always mean "Give me a default" and <code>""</code> to mean "Give me PersistentVolume that also has <code>""</code> StorageClass name". 
+
+In other words we need to distinguish two main cases, either default StorageClass is not present in which case there's no behavior change or where there is a default in which case PVCs with <code>nil</code> won't bind to PVs anymore and instead will be updated to have the StorageClass.
```

Comment:
Does this sound accurate because this is what I understood?
```suggestion
Summarizing the above, we've changed the semantics of <code>nil<code> so that its behavior depends on the presence or absence of a definition of default StorageClass.
```